### PR TITLE
Add more smart pointers to RenderSearchField and SearchPopup

### PR DESCRIPTION
**Pull request by abigailfox**
#### 2fa09212b231cb8c79895fcde70eebf92835601d
<pre>
Add more smart pointers to RenderSearchField and SearchPopup
<a href="https://bugs.webkit.org/show_bug.cgi?id=271504">https://bugs.webkit.org/show_bug.cgi?id=271504</a>
<a href="https://rdar.apple.com/125268281">rdar://125268281</a>

Reviewed by Ryosuke Niwa and Chris Dumez.

Add more smart pointers to RenderSearchField and modify supporting files and functions as needed

* Source/WebCore/loader/EmptyClients.cpp:
* Source/WebCore/platform/SearchPopupMenu.h:
* Source/WebCore/rendering/RenderSearchField.cpp:
(WebCore::RenderSearchField::willBeDestroyed):
(WebCore::RenderSearchField::resultsButtonElement const):
(WebCore::RenderSearchField::cancelButtonElement const):
(WebCore::RenderSearchField::addSearchResult):
(WebCore::RenderSearchField::showPopup):
(WebCore::RenderSearchField::hidePopup):
(WebCore::RenderSearchField::computeControlLogicalHeight const):
(WebCore::RenderSearchField::updateFromElement):
(WebCore::RenderSearchField::autosaveName const):
(WebCore::RenderSearchField::valueChanged):
(WebCore::RenderSearchField::clientPaddingLeft const):
(WebCore::RenderSearchField::clientPaddingRight const):
(WebCore::RenderSearchField::setTextFromItem):
(WebCore::RenderSearchField::createScrollbar):
* Source/WebKit/WebProcess/WebCoreSupport/WebSearchPopupMenu.cpp:
(WebKit::WebSearchPopupMenu::protectedPopupMenu):
* Source/WebKit/WebProcess/WebCoreSupport/WebSearchPopupMenu.h:
* Source/WebKitLegacy/ios/WebCoreSupport/SearchPopupMenuIOS.cpp:
(SearchPopupMenuIOS::protectedPopupMenu):
* Source/WebKitLegacy/ios/WebCoreSupport/SearchPopupMenuIOS.h:
* Source/WebKitLegacy/mac/WebCoreSupport/SearchPopupMenuMac.h:
* Source/WebKitLegacy/mac/WebCoreSupport/SearchPopupMenuMac.mm:
(SearchPopupMenuMac::protectedPopupMenu):

Canonical link: <a href="https://commits.webkit.org/276849@main">https://commits.webkit.org/276849@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/faa838a71003dc988f11703bc110002480bf5e34

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/45760 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/24887 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/48341 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/48430 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/41795 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/48066 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/29169 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/22285 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/37467 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/46338 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/21967 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39475 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18624 "Found 1 new API test failure: /WPE/TestResources:/webkit/WebKitWebView/sync-request-on-max-conns (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/19367 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/40566 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/3804 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/42071 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/40908 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/50185 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/20751 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/17263 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/44583 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/22056 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/43443 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10183 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/22415 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/21745 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/loader/EmptyClients.cpp [hunk 1]
```diff
@@ -471,6 +471,7 @@ class EmptySearchPopupMenu : public SearchPopupMenu {
 
 private:
     PopupMenu* popupMenu() final { return m_popup.ptr(); }
+    RefPtr<PopupMenu> protectedPopupMenu() final { return m_popup.ptr(); }
```
Comment (cdumez) [inline Source/WebCore/loader/EmptyClients.cpp]: FYI, using `m_popup.copyRef()` would be slightly more efficient. It avoids the RefPtr constructor having the null check the pointer, since we know it is non null.

But also, we don't need this override function at all.

File: Source/WebCore/platform/SearchPopupMenu.h [hunk 2]
```diff
@@ -40,6 +40,7 @@ class SearchPopupMenu : public RefCounted<SearchPopupMenu> {
 public:
     virtual ~SearchPopupMenu() = default;
     virtual PopupMenu* popupMenu() = 0;
+    virtual RefPtr<PopupMenu> protectedPopupMenu() = 0;
```
Comment (cdumez) [inline Source/WebCore/platform/SearchPopupMenu.h]: This doesn't need to be virtual, you can do:
```
RefPtr<PopupMenu> protectedPopupMenu() { return popupMenu(); }
```

File: Source/WebCore/rendering/RenderSearchField.cpp [hunk 3]
```diff
@@ -82,20 +82,23 @@ void RenderSearchField::willBeDestroyed()
 
 inline HTMLElement* RenderSearchField::resultsButtonElement() const
 {
-    return inputElement().resultsButtonElement();
```
Comment (cdumez) [inline Source/WebCore/rendering/RenderSearchField.cpp:85]: We should introduce a `protectedInputElement()` and use it here, to avoid creating a new local variable.

File: Source/WebCore/rendering/RenderSearchField.cpp [hunk 4]
```diff
@@ -82,20 +82,23 @@ void RenderSearchField::willBeDestroyed()
...
+    Ref input = inputElement();
+    return input->resultsButtonElement();
 }
 
 inline HTMLElement* RenderSearchField::cancelButtonElement() const
 {
-    return inputElement().cancelButtonElement();
+    Ref input = inputElement();
```
Comment (cdumez) [inline Source/WebCore/rendering/RenderSearchField.cpp]: ditto

File: Source/WebCore/rendering/RenderSearchField.cpp [hunk 5]
```diff
@@ -146,24 +149,25 @@ void RenderSearchField::showPopup()
     FloatPoint absTopLeft = localToAbsolute(FloatPoint(), UseTransforms);
     IntRect absBounds = absoluteBoundingBoxRectIgnoringTransforms();
     absBounds.setLocation(roundedIntPoint(absTopLeft));
-    m_searchPopup->popupMenu()->show(absBounds, &view().frameView(), -1);
+    RefPtr popupMenu = m_searchPopup->protectedPopupMenu();
```
Comment (cdumez) [inline Source/WebCore/rendering/RenderSearchField.cpp]: We're failing to protect m_searchPopup here, this should be something like:
```
protectedSearchPopup()->protectedPopupMenu()->show(absBounds, &view().frameView(), -1);
```

File: Source/WebCore/rendering/RenderSearchField.cpp [hunk 6]
```diff
@@ -191,8 +195,10 @@ void RenderSearchField::updateFromElement()
     if (cancelButtonElement())
         updateCancelButtonVisibility();
 
-    if (m_searchPopupIsVisible)
-        m_searchPopup->popupMenu()->updateFromElement();
+    if (m_searchPopupIsVisible) {
+        RefPtr popupMenu = m_searchPopup->popupMenu();
+        popupMenu->updateFromElement();
```
Comment (cdumez) [inline Source/WebCore/rendering/RenderSearchField.cpp]: `protectedSearchPopup()->protectedPopupMenu()->updateFromElement();`. We shouldn't need a new local variable.

File: Source/WebCore/rendering/RenderSearchField.cpp [hunk 7]
```diff
@@ -218,7 +224,8 @@ Visibility RenderSearchField::visibilityForCancelButton() const
 
 const AtomString& RenderSearchField::autosaveName() const
 {
-    return inputElement().attributeWithoutSynchronization(nameAttr);
+    Ref input = inputElement();
+    return input->attributeWithoutSynchronization(nameAttr);
```
Comment (cdumez) [inline Source/WebCore/rendering/RenderSearchField.cpp]: `return protectedInputElement()->attributeWithoutSynchronization(nameAttr);`

We shouldn't need a new local variable.

File: Source/WebCore/rendering/RenderSearchField.cpp [hunk 8]
```diff
@@ -82,20 +82,23 @@ void RenderSearchField::willBeDestroyed()
 
 inline HTMLElement* RenderSearchField::resultsButtonElement() const
 {
-    return inputElement().resultsButtonElement();
+    Ref input = inputElement();
```
Comment (rniwa) [inline Source/WebCore/rendering/RenderSearchField.cpp]: Maybe we should add protectedInputElement()?

File: Source/WebCore/rendering/RenderSearchField.cpp [hunk 9]
```diff
@@ -310,16 +318,19 @@ int RenderSearchField::clientInsetRight() const
...
 }
 
 LayoutUnit RenderSearchField::clientPaddingRight() const
 {
     LayoutUnit padding = paddingRight();
-    if (RenderBox* containerBox = containerElement() ? containerElement()->renderBox() : 0) {
-        if (RenderBox* innerBlockBox = innerBlockElement() ? innerBlockElement()->renderBox() : 0)
+    RefPtr container = containerElement();
+    RefPtr innerBlock = innerBlockElement();
+    if (RenderBox* containerBox = container ? container->renderBox() : 0) {
```
Comment (cdumez) [inline Source/WebCore/rendering/RenderSearchField.cpp]: Please use `nullptr` instead of `0` since you're touching this line.

File: Source/WebCore/rendering/RenderSearchField.cpp [hunk 10]
```diff
@@ -310,16 +318,19 @@ int RenderSearchField::clientInsetRight() const
...
+        if (RenderBox* innerBlockBox = innerBlock ? innerBlock->renderBox() : 0)
```
Comment (cdumez) [inline Source/WebCore/rendering/RenderSearchField.cpp]: ditto.

File: Source/WebCore/rendering/RenderSearchField.cpp (hunk 9 above)
Comment (cdumez) [reply Source/WebCore/rendering/RenderSearchField.cpp]: Also using `auto*` would be good.

File: Source/WebCore/rendering/RenderSearchField.cpp [hunk 11]
```diff
@@ -362,7 +373,8 @@ bool RenderSearchField::itemIsSelected(unsigned) const
 
 void RenderSearchField::setTextFromItem(unsigned listIndex)
 {
-    inputElement().setValue(itemText(listIndex));
+    Ref input = inputElement();
+    input->setValue(itemText(listIndex));
```
Comment (cdumez) [inline Source/WebCore/rendering/RenderSearchField.cpp]: `protectedInputElement()->setValue(itemText(listIndex));`

File: Source/WebCore/rendering/RenderSearchField.cpp [hunk 12]
```diff
@@ -378,8 +390,9 @@ HostWindow* RenderSearchField::hostWindow() const
 Ref<Scrollbar> RenderSearchField::createScrollbar(ScrollableArea& scrollableArea, ScrollbarOrientation orientation, ScrollbarWidth widthStyle)
 {
     bool usesLegacyScrollbarStyle = style().usesLegacyScrollbarStyle();
+    RefPtr input = &inputElement();
```
Comment (rniwa) [inline Source/WebCore/rendering/RenderSearchField.cpp]: Why not Ref?

File: Source/WebCore/rendering/RenderSearchField.cpp [hunk 13]
```diff
@@ -378,8 +390,9 @@ HostWindow* RenderSearchField::hostWindow() const
...
     if (usesLegacyScrollbarStyle)
-        return RenderScrollbar::createCustomScrollbar(scrollableArea, orientation, &inputElement());
+        return RenderScrollbar::createCustomScrollbar(scrollableArea, orientation, input.get());
```
Comment (rniwa) [inline Source/WebCore/rendering/RenderSearchField.cpp]: You can call input.ptr() once you've changed the above to use Ref instead.

Comment (cdumez) [inline Source/WebCore/rendering/RenderSearchField.cpp]: `protectedInputElement().ptr()`

**Review by rniwa — APPROVED** [review]

File: Source/WebKit/WebProcess/WebCoreSupport/WebSearchPopupMenu.cpp [hunk 14]
```diff
@@ -47,6 +47,11 @@ PopupMenu* WebSearchPopupMenu::popupMenu()
     return m_popup.get();
 }
 
+RefPtr<PopupMenu> WebSearchPopupMenu::protectedPopupMenu()
+{
+    return m_popup.get();
```
Comment (cdumez) [inline Source/WebKit/WebProcess/WebCoreSupport/WebSearchPopupMenu.cpp]: `.get()` is not needed. Also, the function can probably be inlined in the header given that it is so trivial.

File: Source/WebKitLegacy/ios/WebCoreSupport/SearchPopupMenuIOS.cpp [hunk 15]
```diff
@@ -35,6 +35,11 @@ PopupMenu* SearchPopupMenuIOS::popupMenu()
     return m_popup.ptr();
 }
 
+RefPtr<PopupMenu> SearchPopupMenuIOS::protectedPopupMenu()
+{
+    return m_popup.ptr();
```
Comment (cdumez) [inline Source/WebKitLegacy/ios/WebCoreSupport/SearchPopupMenuIOS.cpp]: ditto.

File: Source/WebKitLegacy/mac/WebCoreSupport/SearchPopupMenuMac.mm [hunk 16]
```diff
@@ -49,6 +49,11 @@ static String defaultSearchFieldRecentSearchesStorageDirectory()
     return m_popup.get();
 }
 
+RefPtr<WebCore::PopupMenu> SearchPopupMenuMac::protectedPopupMenu()
+{
+    return m_popup.get();
```
Comment (cdumez) [inline Source/WebKitLegacy/mac/WebCoreSupport/SearchPopupMenuMac.mm]: ditto.

File: Source/WebCore/rendering/RenderSearchField.cpp (hunk 12 above)
Comment (abigailfox) [reply Source/WebCore/rendering/RenderSearchField.cpp]: I can do 
```    
Ref input = inputElement(); 
if (usesLegacyScrollbarStyle)
        return RenderScrollbar::createCustomScrollbar(scrollableArea, orientation, input.ptr());
```
if that's preferred. `createCustomScrollbar` needed a pointer

Comment (cdumez) [reply Source/WebCore/rendering/RenderSearchField.cpp]: Yes, that's definitely preferred. No reason to use a nullable pointer type (RefPtr) when you can use a non-nullable one (Ref).

File: Source/WebCore/rendering/RenderSearchField.h [hunk 17]
```diff
@@ -79,6 +79,7 @@ class RenderSearchField final : public RenderTextControlSingleLine, private Popu
     FontSelector* fontSelector() const override;
     HostWindow* hostWindow() const override;
     Ref<Scrollbar> createScrollbar(ScrollableArea&, ScrollbarOrientation, ScrollbarWidth) override;
+    RefPtr<SearchPopupMenu> protectedSearchPopup() { return m_searchPopup; };
```
Comment (cdumez) [inline Source/WebCore/rendering/RenderSearchField.h]: Function should probably be `const`.

File: Source/WebKitLegacy/mac/WebCoreSupport/SearchPopupMenuMac.h [hunk 18]
```diff
@@ -35,6 +35,7 @@ class SearchPopupMenuMac : public WebCore::SearchPopupMenu {
     ~SearchPopupMenuMac();
 
     WebCore::PopupMenu* popupMenu() override;
+    RefPtr<WebCore::PopupMenu> protectedPopupMenu() { return popupMenu(); }
```
Comment (cdumez) [inline Source/WebKitLegacy/mac/WebCoreSupport/SearchPopupMenuMac.h]: I don't think we need this?

File: Source/WebKitLegacy/ios/WebCoreSupport/SearchPopupMenuIOS.h [hunk 19]
```diff
@@ -33,6 +33,7 @@ class SearchPopupMenuIOS : public WebCore::SearchPopupMenu {
     SearchPopupMenuIOS(WebCore::PopupMenuClient*);
 
     WebCore::PopupMenu* popupMenu() override;
+    RefPtr<WebCore::PopupMenu> protectedPopupMenu() { return popupMenu(); }
```
Comment (cdumez) [inline Source/WebKitLegacy/ios/WebCoreSupport/SearchPopupMenuIOS.h]: I don't think we need this?

File: Source/WebKit/WebProcess/WebCoreSupport/WebSearchPopupMenu.h [hunk 20]
```diff
@@ -31,6 +31,7 @@ class WebSearchPopupMenu : public WebCore::SearchPopupMenu {
     static Ref<WebSearchPopupMenu> create(WebPage*, WebCore::PopupMenuClient*);
 
     WebCore::PopupMenu* popupMenu() override;
+    RefPtr<WebCore::PopupMenu> protectedPopupMenu() { return popupMenu(); }
```
Comment (cdumez) [inline Source/WebKit/WebProcess/WebCoreSupport/WebSearchPopupMenu.h]: I don't think we need this?

Comment (abigailfox) [reply Source/WebKit/WebProcess/WebCoreSupport/WebSearchPopupMenu.h]: Sorry, originally it was complaining about these, but I think it was a consequence of the way I wrote it previously. It seems fine without them now!

File: Source/WebCore/rendering/RenderSearchField.cpp [hunk 21]
```diff
@@ -73,7 +73,7 @@ RenderSearchField::~RenderSearchField()
 void RenderSearchField::willBeDestroyed()
 {
     if (m_searchPopup) {
```
Comment (cdumez) [inline Source/WebCore/rendering/RenderSearchField.cpp]: The way we usually write this is:
```
if (RefPtr searchPopup = std::exchange(m_searchPopup, nullptr))
    searchPopup->popupMenu()->disconnectClient();
```

File: Source/WebCore/rendering/RenderSearchField.cpp [hunk 22]
```diff
@@ -82,20 +82,20 @@ void RenderSearchField::willBeDestroyed()
...
 inline HTMLElement* RenderSearchField::cancelButtonElement() const
 {
-    return inputElement().cancelButtonElement();
+    return protectedInputElement()->cancelButtonElement();
 }
 
 void RenderSearchField::addSearchResult()
 {
-    if (inputElement().maxResults() <= 0)
+    if (protectedInputElement()->maxResults() <= 0)
```
Comment (cdumez) [inline Source/WebCore/rendering/RenderSearchField.cpp]: You need a local variable, we don't want to call `protectedInputElement()` several times in the same scope, this is bad for perf.

```
Ref inputElement = this->inputElement();
// Then use inputElement in this function.
```

File: Source/WebCore/rendering/RenderSearchField.cpp [hunk 23]
```diff
@@ -310,16 +311,19 @@ int RenderSearchField::clientInsetRight() const
 LayoutUnit RenderSearchField::clientPaddingLeft() const
 {
     LayoutUnit padding = paddingLeft();
-    if (RenderBox* box = innerBlockElement() ? innerBlockElement()->renderBox() : 0)
+    RefPtr innerBlock = innerBlockElement();
+    if (RenderBox* box = innerBlock ? innerBlock->renderBox() : 0)
```
Comment (cdumez) [inline Source/WebCore/rendering/RenderSearchField.cpp]: `0` -> `nullptr`

Comment (cdumez) [reply Source/WebCore/rendering/RenderSearchField.cpp]: also `RenderBox*` -> `auto*`

File: Source/WebCore/rendering/RenderSearchField.cpp [hunk 24]
```diff
@@ -310,16 +311,19 @@ int RenderSearchField::clientInsetRight() const
...
     return padding;
 }
 
 LayoutUnit RenderSearchField::clientPaddingRight() const
 {
     LayoutUnit padding = paddingRight();
-    if (RenderBox* containerBox = containerElement() ? containerElement()->renderBox() : 0) {
-        if (RenderBox* innerBlockBox = innerBlockElement() ? innerBlockElement()->renderBox() : 0)
+    RefPtr container = containerElement();
+    RefPtr innerBlock = innerBlockElement();
```
Comment (cdumez) [inline Source/WebCore/rendering/RenderSearchField.cpp]: This should be moved inside the `if` scope, no reason to ref this if we don't end up using it.

Comment (abigailfox) [reply Source/WebCore/rendering/RenderSearchField.cpp]: ~Don't we need to have them exist in order to execute the if statement?~ Never mind, I think I misunderstood

Comment (cdumez) [reply Source/WebCore/rendering/RenderSearchField.cpp]: Is this clearer?
```
RefPtr container = containerElement();
if (auto* containerBox = container ? container->renderBox() : nullptr) {
    RefPtr innerBlock = innerBlockElement();
    if (auto* innerBlockBox = innerBlock ? innerBlock->renderBox() : nullptr)
```

File: Source/WebCore/rendering/RenderSearchField.cpp [hunk 25]
```diff
@@ -108,10 +107,10 @@ void RenderSearchField::addSearchResult()
 
     RecentSearch recentSearch = { value, WallTime::now() };
     m_recentSearches.insert(0, recentSearch);
-    while (static_cast<int>(m_recentSearches.size()) > inputElement().maxResults())
+    while (static_cast<int>(m_recentSearches.size()) > inputElement->maxResults())
         m_recentSearches.removeLast();
 
-    const AtomString& name = autosaveName();
+    AtomString name = autosaveName();
```
Comment (cdumez) [inline Source/WebCore/rendering/RenderSearchField.cpp:113]: Why this change?

**Review by cdumez — APPROVED** [review]
Looks good now, except for one minor AtomString change that seems unrelated to this patch and should probably be reverted.
